src/include/daos_srv/dtx_srv.h

```diff
@@ -109,47 +134,34 @@ enum dtx_status {
 	DTX_ST_COMMITTED	= 2,
 };
 
-struct dtx_exec_arg;
-
-struct dtx_exec_shard_arg {
-	struct daos_shard_tgt		*exec_shard_tgt;
-	struct dtx_exec_arg		*exec_arg;
-	struct dtx_conflict_entry	 exec_dce;
-	int				 exec_shard_rc;
-};
+int
+dtx_leader_begin(struct dtx_id *dti, daos_unit_oid_t *oid, daos_handle_t coh,
+		 daos_epoch_t epoch, uint64_t dkey_hash, uint32_t pm_ver,
+		 uint32_t intent, struct daos_shard_tgt *tgts, int tgts_cnt,
+		 struct dtx_leader_handle *dlh);
+int
+dtx_leader_end(struct dtx_leader_handle *dlh, struct ds_cont_hdl *cont_hdl,
+	       struct ds_cont_child *cont, int result);
 
-typedef void (*dtx_exec_shard_comp_cb_t)(struct dtx_exec_shard_arg *arg,
-					 int rc);
-typedef int (*dtx_exec_shard_func_t)(struct dtx_handle *dth, void *arg, int idx,
-				     dtx_exec_shard_comp_cb_t comp_cb,
-				     struct dtx_exec_shard_arg *comp_cb_arg);
-struct dtx_exec_arg {
-	dtx_exec_shard_func_t	exec_func;
-	void			*exec_func_arg;
-	struct dtx_handle	*dth;
-	ABT_future		future;
-	uint32_t		shard_cnt;
-	int			exec_result;
-	struct dtx_exec_shard_arg exec_shards_args[0];
-};
+typedef void (*dtx_sub_comp_cb_t)(struct dtx_leader_handle *dlh, int idx,
+				  int rc);
+typedef int (*dtx_sub_func_t)(struct dtx_leader_handle *dlh, void *arg, int idx,
+			      dtx_sub_comp_cb_t comp_cb);
 
 int dtx_resync(daos_handle_t po_hdl, uuid_t po_uuid, uuid_t co_uuid,
 	       uint32_t ver, bool block);
-
-int dtx_begin(struct dtx_id *dti, daos_unit_oid_t *oid, daos_handle_t coh,
-	      daos_epoch_t epoch, uint64_t dkey_hash,
-	      struct dtx_conflict_entry *conflict, struct dtx_id *dti_cos,
-	      int dti_cos_count, uint32_t pm_ver, uint32_t intent, bool leader,
-	      struct dtx_handle **dth);
-
-int dtx_end(struct dtx_handle *dth, struct ds_cont_hdl *cont_hdl,
-	    struct ds_cont_child *cont, int result);
-
-int dtx_conflict(daos_handle_t coh, struct dtx_handle *dth, uuid_t po_uuid,
-		 uuid_t co_uuid, struct dtx_conflict_entry *dces, int count,
-		 uint32_t version);
-int dtx_exec_ops(struct daos_shard_tgt *shard_tgts, int tgts_cnt,
-		 struct dtx_handle *dth, dtx_exec_shard_func_t exec_func,
+int
+dtx_begin(struct dtx_id *dti, daos_unit_oid_t *oid, daos_handle_t coh,
+	  daos_epoch_t epoch, uint64_t dkey_hash,
+	  struct dtx_conflict_entry *conflict, struct dtx_id *dti_cos,
+	  int dti_cos_cnt, uint32_t pm_ver, uint32_t intent,
+	  struct dtx_handle *dth);
+int
+dtx_end(struct dtx_handle *dth, struct ds_cont_hdl *cont_hdl,
+	struct ds_cont_child *cont, struct dtx_id *dti_cos, int dti_cos_cnt,
```

Comment:
The @dti_cos and @dti_cos_cnt seems redundant since the @dth has members for that.

Reply:
right